# Week 2
Coursework for CMEE week 2.
## Topics:
* Python

## Contents
### [Code](https://github.com/SamT123/CMEECoursework/tree/master/Week2/Code)
* basic_io1.py
	* Demonstrate reading and writing of text files
* basic_io2.py
	* Demonstrate reading and writing of text files
* basic_io3.py
	* Demonstrate reading and writing of text files
* basic_csv.py
	* Demonstrate reading and writing of csv files
* cfexercises1.py
	* Demonstrate conditional statements, iteration and recursion as methods of control flow
	* This script has been converted into a program in a Practical exercise.
* cfexercises2.py
	* Demonstrate iteration
* loops.py
	* Demonstrate iteration and conditional statements
* oaks.py
	* Demonstrate list comprehensions
* scope.py
	* Demonstrate variable scope and local environments
* boilerplate.py
	* Demonstrate how a python program is composed
* using_name.py
	* Demonstrate how control flow can be used to alter program behaviour based on how the module is called
* sysargv.py
	* Demonstrate how command line arguemnts can be accessed in a python script
* control_flow.py
	* A collection of functions in a python program to demonstrate program/script differences
* lc1.py
	* List comprehension exercises on bird weight data
* lc2.py
	* List comprehension exercises on rainfall data
* tuple.py
	* List comprehension exercises on bird weight data
* dictionary.py
	* Dictionary comperehesnion exercise on taxonomic data
* test_control_flow.py
	* Demonstrate doctests
* debugme.py
	* Script with divide by zero error to demonstrate debugging methods
* align_seqs.py
	* Finds best alignment (without indels) between two sequences provided in ../Data/sequences.csv
	* Best alignment is saved to ../Results/best_align.txt
	* Example usage:
	`python3 align_seqs_better.py`
* align_seqs_fasta.py
	* Finds best alignment (without indels) between sequences provided as command line arguments
	* If no arguments are passed, defaults are used
	* Best alignment is saved to ../Results/best_align_fasta.txt
	* Example usage:
	`python3 align_seqs_fasta.py ../Data/fasta/407228326.fasta ../Data/fasta/407228412.fasta`
* align_seqs_better.py
	* Finds best alignment (without indels) between two sequences provided in ../Data/sequences.csv
	* If multiple best alignments are present, these are all provided
	* Best alignments are saved to ../Results/best_align_better.txt
	* Example usage:
	`python3 align_seqs_better.py`
* oaks_debugme.py
	* Program which determines whether a species falls within the oak genus, Quercus.





## [Data](https://github.com/SamT123/CMEECoursework/tree/master/Week2/Data)
* testcsv.csv
	* csv file with bird species information
* fasta
	* fasta files for alignment by align_seqs_fasta.py
* TestOaksData.csv
	* species names to categorise
* bodymass.csv
	* bird bodymass and species name information
* sequences.csv
	* DNA sequences to align
## [Results](https://github.com/SamT123/CMEECoursework/tree/master/Week2/Results)
